File: .tmp_py4013/batch_08_dump/2949_count_beautiful_substrings_ii.js
// LeetCode 2949 - Count Beautiful Substrings II
// https://leetcode.com/problems/count-beautiful-substrings-ii/

function isVowel(c) {
    return c === 'a' || c === 'e' || c === 'i' || c === 'o' || c === 'u';
}
var beautifulSubstrings = function(s, k) {
    let x = 1;
    while ((x * x) % k !== 0) x++;
    const freq = new Map();
    freq.set(0, 1);
    let bal = 0, vowels = 0, ans = 0;
    for (let i = 0; i < s.length; i++) {
        const ch = s[i];
        if (isVowel(ch)) { bal++; vowels++; }
        else bal--;
        const kk = (BigInt(bal) << 32n) | BigInt(vowels % x);
        const key = kk.toString();
        const f = freq.get(key) || 0;
        ans += f;
        freq.set(key, f + 1);
    }
    return ans;
};

File: .tmp_rs4013/rust_check13/2949_count_beautiful_substrings_ii.rs
#![allow(dead_code, unused_variables, unused_mut, unused_assignments, unused_imports)]
struct Solution;
// LeetCode 2949 - Count Beautiful Substrings II
// https://leetcode.com/problems/count-beautiful-substrings-ii/

use std::collections::HashMap;

impl Solution {
    pub fn beautiful_substrings(s: String, k: i32) -> i64 {
        fn is_vowel(c: u8) -> bool {
            matches!(c, b'a' | b'e' | b'i' | b'o' | b'u')
        }
        let mut x = 1;
        while (x * x) % k != 0 {
            x += 1;
        }
        let mut freq: HashMap<(i32, i32), i64> = HashMap::new();
        freq.insert((0, 0), 1);
        let mut bal = 0i32;
        let mut vowels = 0i32;
        let mut ans = 0i64;
        for &ch in s.as_bytes() {
            if is_vowel(ch) {
                bal += 1;
                vowels += 1;
            } else {
                bal -= 1;
            }
            let kk = (bal, vowels % x);
            ans += *freq.get(&kk).unwrap_or(&0);
            *freq.entry(kk).or_insert(0) += 1;
        }
        ans
    }
}
